# Milestone 6: Conditional Probability, Independence, Bayes’ Rule, and Probability Distributions
 Sajib Chowdhury 
 ID : 242014141

Objective: Apply concepts of Conditional Probability, Independence, Bayes’ Rule, and the Normal Probability Distribution to a real-world dataset.

**Instructions:** Complete the code and analysis in the sections below. Replace the placeholder values and comments with your specific variable names, events, calculations, and interpretations.

## A. Setup and Data Loading

In [25]:
# Import necessary libraries
import pandas as pd
import numpy as np
from scipy.stats import norm
import matplotlib.pyplot as plt
import seaborn as sns

# Set a consistent visual style
sns.set_style('whitegrid')

# Load your dataset (Replace 'your_data.csv' with your actual file name)
df = pd.read_csv('academic Stress level -Dataset.csv')

# Display the first few rows and info to confirm load
print("Data loaded successfully:")
print(df.head().to_markdown(index=False, numalign="left", stralign="left"))
df.info()

# --- Data Preprocessing (Optional: Add your cleaning/filtering steps here) ---

Data loaded successfully:
| Timestamp           | Your Academic Stage   | Peer pressure   | Academic pressure from your home   | Study Environment   | What coping strategy you use as a student?         | Do you have any bad habits like smoking, drinking on a daily basis?   | What would you rate the academic  competition in your student life   | Rate your academic stress index    |
|:--------------------|:----------------------|:----------------|:-----------------------------------|:--------------------|:---------------------------------------------------|:----------------------------------------------------------------------|:---------------------------------------------------------------------|:-----------------------------------|
| 24/07/2025 22:05:39 | undergraduate         | 4               | 5                                  | Noisy               | Analyze the situation and handle it with intellect | No                                                                    | 3       

# C. Task 1: Define Events

Event,Description (in words),Mathematical Notation

A  .     Student has High Academic Stress Index (Rating ≥4).,A: Academic Stress Index≥4

B   .    Student is an Undergraduate.,B: Academic Stage=’undergraduate’

C    .   Student has Low Academic Stress Index (Rating ≤2).,C: Academic Stress Index≤2

In [26]:
# --- Define helper functions/filters for the events (A and B) based on your DataFrame ---

# Define Event A: High Academic Stress Index (Rating >= 4)
column_A = 'Rate your academic stress index '
threshold_A = 4
filter_A = df[column_A] >= threshold_A
count_A = len(df[filter_A])

# Define Event B: Undergraduate Student
column_B = 'Your Academic Stage'
category_B = 'undergraduate'
filter_B = df[column_B] == category_B
count_B = len(df[filter_B])

# Event A AND B (Intersection: High Stress AND Undergraduate)
filter_A_and_B = filter_A & filter_B
count_A_and_B = len(df[filter_A_and_B])

N = len(df) # Total number of observations

print(f"Total Observations (N): {N}")
print(f"Count(A): {count_A}")
print(f"Count(B): {count_B}")
print(f"Count(A ∩ B): {count_A_and_B}")

Total Observations (N): 140
Count(A): 89
Count(B): 100
Count(A ∩ B): 62


## D. Task 2: Conditional Probability

### 💡 Knowledge Point: Conditional Probability

The conditional probability of event **A occurring, given that event B has already occurred**, is calculated as:

$$
P(A \mid B) = \frac{P(A \cap B)}{P(B)}
$$

Where:
* $P(A \cap B)$ is the probability of both A and B occurring (**Joint Probability**).
* $P(B)$ is the probability of B occurring (**Marginal Probability**).

Compute $P(A)$, $P(B)$, and $P(A \mid B)$ for your chosen events A and B.


solution : D. Task 2: Conditional ProbabilityComputed ProbabilitiesThe probabilities are computed using the counts from Task 1:

N = 140;

Count(A) = 89(High Stress Index>= 4);

Count(B) = 100(Undergraduate Student);

Count(A \cap B) = 62(High Stress AND Undergraduate).


In [27]:
# Counts from Task 1
N = 140
count_A = 89 # Event A: High Academic Stress Index (Rating >= 4)
count_B = 100 # Event B: Undergraduate Student
count_A_and_B = 62 # Event A ∩ B: High Stress AND Undergraduate

# 1. Compute P(A) and P(B)
P_A = count_A / N
P_B = count_B / N

# 2. Compute P(A ∩ B)
P_A_and_B = count_A_and_B / N

# 3. Compute P(A | B) using the formula
P_A_given_B = P_A_and_B / P_B

print(f"P(A) = {P_A:.4f}")
print(f"P(B) = {P_B:.4f}")
print(f"P(A ∩ B) = {P_A_and_B:.4f}")
print(f"P(A | B) = {P_A_given_B:.4f}")

P(A) = 0.6357
P(B) = 0.7143
P(A ∩ B) = 0.4429
P(A | B) = 0.6200


###Interpretation

1.P(A) = 0.6357$ (Marginal Probability)

   Interpretation: The overall probability of a student having a High Academic Stress Index (>= 4) is approximately 63.57%

2.P(B) = 0.7143(Marginal Probability)

  Interpretation: The overall probability of a student being an Undergraduate is approximately 71.43%

3.P(A \ B) = 0.6200(Conditional Probability)

    Interpretation: The probability that a student has a High  Academic Stress Index (>= 4) given that they are an Undergraduate is $62.00\%.

Comparison:   The conditional probability P(A \ B) = 0.6200  is   slightly lower than the marginal probability P(A) = 0.6357$. This suggests that being an Undergraduate (Event B) slightly decreases the likelihood of having a High Academic Stress Index (Event A), or at least, the likelihood among Undergraduates is slightly less than the likelihood in the general student population in this dataset.

## E. Task 3: Independence Check

### 💡 Knowledge Point: Independence of Events

Two events $A$ and $B$ are **independent** if the occurrence of one does not affect the probability of the other. Mathematically, this is checked by the **Multiplication Rule for Independent Events**:

$$
P(A \cap B) = P(A) P(B)
$$

If $P(A \cap B) \neq P(A) P(B)$, the events are **dependent**.

Check if events $A$ and $B$ are independent by comparing the joint probability $P(A \cap B)$ with the product $P(A)P(B)$.




In [28]:
# Counts and N from Task 1
N = 140
count_A = 89
count_B = 100
count_A_and_B = 62

# Probabilities
P_A = count_A / N
P_B = count_B / N
P_A_and_B = count_A_and_B / N

# 1. Compute P(A) * P(B)
product_P_A_P_B = P_A * P_B

# 2. Compare P(A ∩ B) with P(A)P(B)
print(f"P(A ∩ B) (Empirical Joint Prob): {P_A_and_B:.4f}")
print(f"P(A) * P(B) (Product of Marginal Probs): {product_P_A_P_B:.4f}")

difference = abs(P_A_and_B - product_P_A_P_B)

# Check for approximate equality (Use a small tolerance for 'independence')
if difference < 0.01:
    print("\nConclusion: P(A ∩ B) is approximately equal to P(A)P(B).")
    print("The events A and B are **Independent**.")
else:
    print("\nConclusion: P(A ∩ B) is NOT equal to P(A)P(B).")
    print("The events A and B are **Dependent**.")

P(A ∩ B) (Empirical Joint Prob): 0.4429
P(A) * P(B) (Product of Marginal Probs): 0.4541

Conclusion: P(A ∩ B) is NOT equal to P(A)P(B).
The events A and B are **Dependent**.


## F. Task 4: Bayes' Rule

### 💡 Knowledge Point: Bayes' Rule

Bayes' Rule allows us to **reverse the conditional probability**. If we know $P(A \mid B)$, we can find $P(B \mid A)$ using the following formula:

$$
P(B \mid A) = \frac{P(A \mid B)\,P(B)}{P(A)}
$$

Use Bayes' Rule to compute $P(B \mid A)$ from the probabilities calculated in Task 2. Then, compare this result with the empirical value $P(B \mid A)$ calculated directly from the data.

In [29]:
# Counts and N from Task 1 (using successful results)
N = 140
count_A = 89
count_B = 100
count_A_and_B = 62

# Re-calculate necessary probabilities from counts for maximum precision
P_A = count_A / N
P_B = count_B / N
P_A_and_B = count_A_and_B / N
P_A_given_B = count_A_and_B / count_B

# 1. Compute P(B | A) using Bayes' Rule
P_B_given_A_Bayes = (P_A_given_B * P_B) / P_A

# 2. Compute P(B | A) empirically from the dataset (P(A ∩ B) / P(A))
P_B_given_A_Empirical = P_A_and_B / P_A

print(f"P(A | B) (from Task 2): {P_A_given_B:.4f}")
print(f"P(A) (from Task 2): {P_A:.4f}")
print(f"P(B) (from Task 2): {P_B:.4f}")
print("\n---")
print(f"P(B | A) via Bayes' Rule: {P_B_given_A_Bayes:.4f}")
print(f"P(B | A) Empirical: {P_B_given_A_Empirical:.4f}")

# The two values should be mathematically identical (within floating point error)
print(f"Difference: {abs(P_B_given_A_Bayes - P_B_given_A_Empirical):.8f}")

P(A | B) (from Task 2): 0.6200
P(A) (from Task 2): 0.6357
P(B) (from Task 2): 0.7143

---
P(B | A) via Bayes' Rule: 0.6966
P(B | A) Empirical: 0.6966
Difference: 0.00000000


## G. Task 5: Probability Distribution (Normal Only)

### G1. Explore a Numerical Variable

### 💡 Knowledge Point: Normal Distribution

The **Normal Distribution** $X \sim N(\mu, \sigma^2)$ is a continuous, bell-shaped distribution defined entirely by its **mean** ($mu$) and **standard deviation** ($sigma$).

The **68-95-99.7 Rule** states that approximately:
* **68%** of data falls within $\mu \pm 1\sigma$.
* **95%** of data falls within $\mu \pm 2\sigma$.

Select a numerical variable (e.g., `Score`) and analyze its fit to the Normal Distribution.

In [30]:
import pandas as pd
import numpy as np
from scipy.stats import norm
import matplotlib.pyplot as plt
import seaborn as sns

# Set a consistent visual style
sns.set_style('whitegrid')

# Load your dataset
df = pd.read_csv('academic Stress level -Dataset.csv')

# Select your numerical variable (Note the trailing space)
numerical_variable = 'Rate your academic stress index '
data_series = df[numerical_variable]

# 1. Compute Mean (μ) and Standard Deviation (σ)
mu = data_series.mean()
sigma = data_series.std()

# FIX: Define var_name_clean to avoid NameError
var_name_clean = numerical_variable.strip()

print(f"Selected Variable: {var_name_clean}")
print(f"Mean (μ): {mu:.2f}")
print(f"Standard Deviation (σ): {sigma:.2f}")

# 2. Plot Histogram and Overlay Normal Curve
plt.figure(figsize=(10, 6))

# Plot the histogram of the data
sns.histplot(data_series, bins=len(data_series.unique()), kde=False, color='skyblue', alpha=0.6, stat="density", discrete=True, label='Data Histogram')

# Generate points for the theoretical Normal Distribution curve
xmin, xmax = plt.xlim()
x = np.linspace(xmin, xmax, 100)
p = norm.pdf(x, mu, sigma) # Probability Density Function (PDF)

# Plot the Normal Distribution curve
plt.plot(x, p, 'k', linewidth=2, label=f'Normal Curve ($\\mu={mu:.2f}, \\sigma={sigma:.2f}$)')

# Use the correctly defined var_name_clean
title = f"Histogram of {var_name_clean} with Normal Distribution Overlay"
plt.title(title)
plt.xlabel(var_name_clean)
plt.ylabel('Density')
plt.legend()
plt.savefig('normal_distribution_overlay.png')
plt.close()

Selected Variable: Rate your academic stress index
Mean (μ): 3.72
Standard Deviation (σ): 1.03


### G2. Normal Probability Questions

### 💡 Knowledge Point: Calculating Normal Probabilities

We use the `scipy.stats.norm` functions for calculating theoretical probabilities:
* **`norm.cdf(x, loc=mu, scale=sigma)`:** Calculates $P(X \le x)$, the **Cumulative Distribution Function**.
* **`norm.sf(x, loc=mu, scale=sigma)`:** Calculates $P(X > x)$, the **Survival Function**.

Use the fitted Normal distribution ($X \sim N(\mu, \sigma^2)$) to compute the required theoretical probabilities.

In [31]:
import pandas as pd
import numpy as np
from scipy.stats import norm

# Load your dataset
df = pd.read_csv('academic Stress level -Dataset.csv')

# Define variables using data (using correct column name with trailing space)
numerical_variable = 'Rate your academic stress index '
data_series = df[numerical_variable]

# Compute Mean (μ) and Standard Deviation (σ)
mu = data_series.mean()
sigma = data_series.std()
var_name_clean = numerical_variable.strip()

# 1. P(X > μ)
P_greater_mu = norm.sf(mu, loc=mu, scale=sigma)

# 2. P(μ - σ < X < μ + σ) - Probability within 1 sigma
lower_1sigma = mu - sigma
upper_1sigma = mu + sigma
P_1sigma = norm.cdf(upper_1sigma, loc=mu, scale=sigma) - norm.cdf(lower_1sigma, loc=mu, scale=sigma)

# 3. P(X < μ - 2σ) - Probability less than 2 sigmas below the mean
lower_2sigma = mu - 2 * sigma
P_less_2sigma = norm.cdf(lower_2sigma, loc=mu, scale=sigma)

print(f"Theoretical Probabilities based on N({mu:.2f}, {sigma**2:.2f}):")
print(f"P(X > μ) = {P_greater_mu:.4f}")
print(f"P({lower_1sigma:.2f} < X < {upper_1sigma:.2f}) [1σ range] = {P_1sigma:.4f}")
print(f"P(X < {lower_2sigma:.2f}) [Less than μ-2σ] = {P_less_2sigma:.4f}")

print("\n--- Interpretation ---")
print(f"Interpretation 1: The theoretical probability of a student having a {var_name_clean} **greater than the mean** is {P_greater_mu:.2%}.")
print(f"Interpretation 2: The theoretical probability of a student having a {var_name_clean} **within one standard deviation** of the mean is {P_1sigma:.2%}.")
print(f"Interpretation 3: The theoretical probability of a student having a {var_name_clean} **less than two standard deviations below the mean** is {P_less_2sigma:.2%}.")

Theoretical Probabilities based on N(3.72, 1.07):
P(X > μ) = 0.5000
P(2.69 < X < 4.75) [1σ range] = 0.6827
P(X < 1.66) [Less than μ-2σ] = 0.0228

--- Interpretation ---
Interpretation 1: The theoretical probability of a student having a Rate your academic stress index **greater than the mean** is 50.00%.
Interpretation 2: The theoretical probability of a student having a Rate your academic stress index **within one standard deviation** of the mean is 68.27%.
Interpretation 3: The theoretical probability of a student having a Rate your academic stress index **less than two standard deviations below the mean** is 2.28%.


### G3. Are Your Data Normally Distributed?

Based on the visualization and descriptive statistics (Mean vs. Median), comment on the normality of the chosen variable.

In [32]:
import pandas as pd

# Load your dataset
df = pd.read_csv('academic Stress level -Dataset.csv')

# Define variables using data (using correct column name with trailing space)
numerical_variable = 'Rate your academic stress index '
data_series = df[numerical_variable]

# Compute Mean (μ) and Median
mu = data_series.mean()
median = data_series.median()

print(f"Mean: {mu:.2f}")
print(f"Median: {median:.2f}")

print("\n--- Comment on Normality ---")
# Analysis based on data:
# Mean (3.72) < Median (4.00) => Left Skew
print("The histogram **does not look** like a perfect bell curve.")
print(f"The Mean ({mu:.2f}) and Median ({median:.2f}) are **far apart**, which suggests **poor** symmetry.")
print("Overall, the data **does not show** a strong indication of following a Normal Distribution. **The distribution is strongly negatively (left) skewed**, clustered heavily around the scores 4 and 5.")

Mean: 3.72
Median: 4.00

--- Comment on Normality ---
The histogram **does not look** like a perfect bell curve.
The Mean (3.72) and Median (4.00) are **far apart**, which suggests **poor** symmetry.
Overall, the data **does not show** a strong indication of following a Normal Distribution. **The distribution is strongly negatively (left) skewed**, clustered heavily around the scores 4 and 5.


## H. Task 6: Reflection

Summarizing the findings from the analysis of the relationship between 'Rate my  academic stress index' (A) and 'my  Academic Stage' (B) and the distribution of stress scores.

### Reflection Summary

1. **Conditional Probability's Effect:**
   
    [Conditional Probability's Effect:
The conditional probability $P(A \mid B) \approx 0.6200$ (High Stress given Undergraduate) was slightly lower than the marginal probability $P(A) \approx 0.6357$ (High Stress overall). This means that being an Undergraduate slightly decreases the likelihood of a student reporting a High Academic Stress Index compared to the general population of students in this dataset. Conditional probability provided a key refinement, suggesting that the problem of high stress is not uniformly distributed across all academic stages.]

2. **Independence of Events:**
    [The events A (High Stress Index $\ge 4$) and B (Undergraduate Student) were found to be Dependent. This was concluded because the empirical joint probability ($P(A \cap B) \approx 0.4429$) did not equal the product of the marginal probabilities ($P(A)P(B) \approx 0.4541$).The dependence implies that the Academic Stage is a real-world factor that influences or is associated with the level of academic stress a student reports. Since they are dependent, knowing a student is an undergraduate provides non-trivial information about their stress rating.]
3. ** Normal Distribution Fit:**
    [Normal Distribution Fit:The numerical variable, 'Rate your academic stress index', did not fit the Normal distribution well. The data exhibited a strong negative (left) skew, clustered heavily at the higher stress scores (4 and 5). This was supported by the Mean ($\mu=3.72$) being lower than the Median ($4.00$) and the visible non-bell shape of the histogram.Non-normality affects many common parametric statistical tests (like t-tests or ANOVA) that assume data is normally distributed.1 Using these tests on highly skewed data can lead to invalid p-values and unreliable conclusions, requiring the use of non-parametric alternatives or data transformations.]
4. **Real-World Application:**
   [Bayes' Rule in Medical Diagnosis:Bayes' Rule is foundational in determining the probability of a disease, given a positive test result (2$P(\text{Disease} \mid \text{Positive Test})$).3For example, imagine a rare disease $D$ ($P(D)$ is low) and a test that is $99\%$ accurate ($P(\text{Positive} \mid D)$ is high). A physician must use Bayes' Rule:$$P(\text{Disease} \mid \text{Positive}) = \frac{P(\text{Positive} \mid \text{Disease}) \times P(\text{Disease})}{P(\text{Positive})}$$Because the prior probability $P(D)$ is often very low, the posterior probability $P(\text{Disease} \mid \text{Positive})$ might still be low even with a highly accurate test. This rule prevents doctors from misinterpreting a positive result for a rare disease as a near-certain diagnosis, guiding further diagnostic steps and treatment decisions.]